chain/blocksync/client.go

```diff
@@ -0,0 +1,445 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return validRes.tipsets, nil
+}
+
+func (client *BlockSync) GetFullTipSet(
+	ctx context.Context,
+	peer peer.ID,
+	tsk types.TipSetKey,
+) (*store.FullTipSet, error) {
+	// TODO: round robin through these peers on error
+
+	req := &Request{
+		Head:    tsk.Cids(),
+		Length:  1,
+		Options: Headers | Messages,
+	}
+
+	validRes, err := client.doRequest(ctx, req, &peer)
+	if err != nil {
+		return nil, err
+	}
+
+	return validRes.toFullTipSets()[0], nil
+	// If `doRequest` didn't fail we are guaranteed to have at least
+	//  *one* tipset here, so it's safe to index directly.
+}
+
+func (client *BlockSync) GetChainMessages(
+	ctx context.Context,
+	head *types.TipSet,
+	length uint64,
+	) ([]*CompactedMessages, error) {
+	ctx, span := trace.StartSpan(ctx, "GetChainMessages")
+	defer span.End()
+
+	req := &Request{
+		Head:    head.Cids(),
+		Length:  length,
+		Options: Messages,
+	}
+
+	validRes, err := client.doRequest(ctx, req, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return validRes.messages, nil
+}
+
+// Send a request to a peer. Write request in the stream and read the
+// response back. We do not do any processing of the request/response
+// here.
+func (client *BlockSync) sendRequestToPeer(
+	ctx context.Context,
+	peer peer.ID,
+	req *Request,
+) (_ *Response, err error) {
+	// Trace code.
+	ctx, span := trace.StartSpan(ctx, "sendRequestToPeer")
+	defer span.End()
+	if span.IsRecordingEvents() {
+		span.AddAttributes(
+			trace.StringAttribute("peer", peer.Pretty()),
+		)
+	}
+	defer func() {
+		if err != nil {
+			if span.IsRecordingEvents() {
+				span.SetStatus(trace.Status{
+					Code:    5,
+					Message: err.Error(),
+				})
+			}
+		}
+	}()
+	// -- TRACE --
+
+	supported, err := client.host.Peerstore().SupportsProtocols(peer, BlockSyncProtocolID)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get protocols for peer: %w", err)
+	}
+	if len(supported) == 0 || supported[0] != BlockSyncProtocolID {
+		return nil, xerrors.Errorf("peer %s does not support protocol %s",
+			peer, BlockSyncProtocolID)
+		// FIXME: `ProtoBook` should support a *single* protocol check that returns
+		//  a bool instead of a list.
+	}
+
+	connectionStart := build.Clock.Now()
+
+	// Open stream to peer.
+	stream, err := client.host.NewStream(
+		inet.WithNoDial(ctx, "should already have connection"),
+		peer,
+		BlockSyncProtocolID)
+	if err != nil {
+		client.RemovePeer(peer)
+		return nil, xerrors.Errorf("failed to open stream to peer: %w", err)
+	}
+
+	// Write request.
+	_ = stream.SetWriteDeadline(time.Now().Add(WRITE_REQ_DEADLINE))
+	if err := cborutil.WriteCborRPC(stream, req); err != nil {
+		_ = stream.SetWriteDeadline(time.Time{})
+		// FIXME: What's the point of setting a blank deadline that won't time out?
+		//  Is this our way of clearing the old one?
+		client.peerTracker.logFailure(peer, build.Clock.Since(connectionStart))
+		// FIXME: Should we also remove peer here?
```

Comment:
i currently assume the peer tracker won't try the peer again unless they are our only other option